apps/prairielearn/src/pages/instructorCourseAdminSharing/instructorCourseAdminSharing.ts

```diff
@@ -0,0 +1,100 @@
+import { Router } from 'express';
+import asyncHandler = require('express-async-handler');
+import error = require('@prairielearn/error');
+import { v4 as uuidv4 } from 'uuid';
+import { CourseSchema } from '../../lib/db-types';
+import { InstructorSharing } from './instructorCourseAdminSharing.html';
+import sqldb = require('@prairielearn/postgres');
+
+const router = Router();
+const sql = sqldb.loadSqlEquiv(__filename);
+
+async function generateSharingId(res) {
+  const newSharingId = uuidv4();
+  return await sqldb.queryZeroOrOneRowAsync(sql.update_sharing_token, {
+    sharing_token: newSharingId,
+    course_id: res.locals.course.id,
+  });
+}
+
+router.get(
+  '/',
+  asyncHandler(async (req, res) => {
+    if (!res.locals.question_sharing_enabled) {
+      throw error.make(403, 'Access denied (feature not available)');
+    }
+
+    const sharingInfo = await sqldb.queryRow(
+      sql.get_course_sharing_info,
+      {
+        course_id: res.locals.course.id,
+      },
+      CourseSchema,
+    );
+
+    if (!sharingInfo.sharing_token) {
+      await generateSharingId(res);
+      res.redirect(req.originalUrl);
+      return;
+    }
+
+    const sharingSets = await sqldb.queryAsync(sql.select_sharing_sets, {
+      course_id: res.locals.course.id,
+    });
+    res.send(
+      InstructorSharing({
+        sharing_name: sharingInfo.sharing_name,
+        sharing_token: sharingInfo.sharing_token,
+        sharing_sets: sharingSets.rows,
+        resLocals: res.locals,
+      }),
+    );
+  }),
+);
+
+router.post(
+  '/',
+  asyncHandler(async (req, res) => {
+    if (!res.locals.authz_data.has_course_permission_own) {
+      throw error.make(403, 'Access denied (must be course owner)');
+    }
+    if (!res.locals.question_sharing_enabled) {
+      throw error.make(403, 'Access denied (feature not available)');
+    }
+
+    if (req.body.__action === 'sharing_token_regenerate') {
+      await generateSharingId(res);
+      res.redirect(req.originalUrl);
+      return;
+    } else if (req.body.__action === 'unsafe_sharing_set_create') {
+      await sqldb.queryZeroOrOneRowAsync(sql.sharing_set_create, {
+        sharing_set_name: req.body.sharing_set_name,
+        course_id: res.locals.course.id,
+      });
+    } else if (req.body.__action === 'unsafe_course_sharing_set_add') {
+      await sqldb.queryZeroOrOneRowAsync(sql.course_sharing_set_add, {
+        sharing_set_id: req.body.sharing_set_id,
```

Comment:
Nice catch! 

While looking into this, I realized that the query was also not disallowing someone adding _their own course_ to their sharing set, which wouldn't be that dangerous but it would be pointless.

I have added validation for both of these cases, along with corresponding tests to ensure correctness.